### DISCUSSION TITLE
Access the learned controller?

### DISCUSSION THREAD

**Discussion by wissamkafa** [hilo-mpc/hilo-mpc#37] [Q&A]
Good evening.
@BrunoMorabito 

I am trying to learn the MPC controlling a specific system (with 3rd-order differential equations).

While training, the loss value decreases to almost zero, but when plotting the results after using the learned controller with 
     scl = SimpleControlLoop(system, ann)  
     scl.run(n_steps)
The results seem very different and do not follow the labels.

1- Should not the loss values indicate the learned controller effectiveness?
2 - How can I access the learned controller (the network) to directly pass values (features) and see the corresponding output, to compare with the original mpc controller?
3 - Did HILO-mpc have been tested with nonlinear models of higher orders?

Original MPC Controller:
![original_MPC](https://github.com/hilo-mpc/hilo-mpc/assets/9342111/34273d21-33d2-4aea-b167-d4bad33a0021)

Learned Controller:
![learned_MPC](https://github.com/hilo-mpc/hilo-mpc/assets/9342111/c07ef1d2-78a3-4ddc-bb49-976c0bf1fc17)

Loss:
![loss](https://github.com/hilo-mpc/hilo-mpc/assets/9342111/4974d5b6-aee8-4eea-9fe1-c30379ff7d74)

**Comment by brunomorampc** [comment]
Hi @wissamkafa,

sorry for the late reply, I missed your message again.
In case you haven't already solved the issue: the low loss tells you that the neural network fits well the training data( maybe it's also overfitting a bit. I would suggest you to plot in logarithm scale on the y axis by the way). Even assuming that your measurements have no noise, when using the controller in a control loop, the values that are passed to the controllers (states) might be new for the neural network (because they never appeared in the training data), this could be the reason why it does not follow the reference.

I would suggest you to give more datapoints which are information-rich, i.e. taken from different experiments.

Regarding how to inspect the MPC, the trained MPC behaves exactly as a normal neural network class: you can call `mpc.predict(x)` to get the input from the state `x`.

**Reply by wissamkafa** [reply]
@brunomorampc 
good day,
Thanks for your reply.
Actually I did not find a solution to that problem yet.

* As you suggest:  "to give more datapoints which are information-rich, i.e. taken from different experiments", 
is that possible and how as the library automatically generate the dataset using:
data_set = system.generate_data('closed_loop', nmpc, steps=n_steps, use_input_as_label=True)
and then train the nn on that generated dataset.
* Also, it seems that there is no "mpc.predict(x)" function.

it is essential to me now to overcome this step and get an acceptable trained network that could replace the MPC.
if possible, can I send the code by email to test and check it?

**Reply by brunomorampc** [reply]
hey @wissamkafa yes you can send me your code

**Reply by wissamkafa** [reply]
@brunomorampc 
good morning
I sent the file-code by email, I hope you have time to check it.

**Reply by brunomorampc** [reply]
Hey @wissamkafa  yes I received it, I'll have time later today.

**Comment by brunomorampc** [comment]
HI Wissam,

Let's go through your issues one at time:

## **Problem with the controller not behaving well**

This is probably due to the fact that you the neural network is seeing very little. When doing 

```
data_set = system.generate_data('closed_loop', nmpc, steps=n_steps, use_input_as_label=True)
```

you are collecting the data of the closed loop system, in other words, the orange lines that you see in these plots

![image](https://github.com/hilo-mpc/hilo-mpc/assets/27070958/dc07ab32-b79e-4e0c-a40e-c575e12fa881)

This is very little information for the neural network, as it might learn well if the system is very close to how the orange line, but as soon as you have (even a veeeery little)  variation the controller unstabilizes the system. One idea to fix is the following. Run several `system.generate_data` with different system and MPC settings where you vary, for example, initial conditions and reference point. For example

```

from hilo_mpc.util.data import DataSet 

list_of_initial_conditions = [ initial_cond_1, initial_cond_2, ..., initial_cond_N]
list_of_set_points = [sp1, sp2, ..., sp_M]
data_set_tot = DataSet(features=[], labels = []) # Generate empty dataset


for ic in list_of_initial_conditions:
     for sp in list_of_set_points:
         ... setup MPC....
         data_set_i = system.generate_data(...)
        data_set_tot.add_data(features=data_set.raw_data[0], labels=data_set.raw_data[1]) # add data to dataset
```

and then train with data_set_tot. One suggestion when testing the learned MPC: start easy, possibly by setting initial conditions and reference to an equilibrium point, in this case the system should stay in the equilibrium point. Make sure that the "areas" close to the equilibrium point are in the training dataset. 

## **How to add external data to the dataset**

The previous example should clarify this. Just create an (empty if you want) dataset and the add features and labels using 'add_data'

## **How to get the trained weights from the trained NN**

I tried but I could not find a way, surely @jpohlodek  can help with that?

**Reply by wissamkafa** [reply]
Dear @brunomorampc 
thanks for your efforts ans helpful instructions.

I will implement them soon, and report the results.

**Comment by wissamkafa** [comment]
@BrunoMorabito 
@brunomorampc 
Good Day

 **Continuing our last discussion about Learning_MPC:**

**Actually the total dataset could not be created in this way (or I have missed something) (because it needs to be of a special structure):**
```
from hilo_mpc.util.data import DataSet 
list_of_initial_conditions = [ initial_cond_1, initial_cond_2, ..., initial_cond_N]
list_of_set_points = [sp1, sp2, ..., sp_M]
data_set_tot = DataSet(features=[], labels = []) # Generate empty dataset


for ic in list_of_initial_conditions:
     for sp in list_of_set_points:
         ... setup MPC....
         data_set_i = system.generate_data(...)
        data_set_tot.add_data(features=data_set.raw_data[0], labels=data_set.raw_data[1]) # add data to dataset

**Instead I had created it using the following method:**
**#First step: create the tot_data_set by running the system on arbitrary set point and initial condition (then we have the correct structure of the dataset)**
#Make controller
nmpc = NMPC(system)
nmpc.quad_stage_cost.add_states()
nmpc.quad_stage_cost.add_inputs()
nmpc.quad_terminal_cost.add_states()
nmpc.set_box_constraints()
#Initial guess
nmpc.set_initial_guess(x_guess=ic, u_guess=0)
#Setup controller
nmpc.setup(options={'print_level': 0, 'ipopt_debugger': False})
data_set_tot = system.generate_data('closed_loop', nmpc, steps=n_steps, use_input_as_label=True)

**#Step 2: create the loop to add more data:**
for ic in list_of_initial_conditions:
     for sp in list_of_set_points:
         ... setup MPC....
         data_set_i = system.generate_data(...)
        data_set_tot.add_data(features=data_set.raw_data[0], labels=data_set.raw_data[1]) # add data to dataset


**Now when training the network on bugger dataset and test as you suggested, the controller does not behave well (it does not stay in the equilibrium point).**

**Actually there is something is confusing here: I couldn't figure out which instruction set the setpoint when testing the learned controller:**

**#Try out the learned controller**
system.reset_solution(keep_initial_conditions=False)
system.set_initial_conditions(ic_ann)
scl = SimpleControlLoop(system, ann)   
scl.run(n_steps)
```
I think the controller (Ann), needs a setpoint ref, and now we are just giving the initial condition and runs with no ref. is that the case, or I missed something?

**Reply by brunomorampc** [reply]
Hi @wissamkafa,

could you share the entire code?

The setpoint is the reference parameter that you pass when defining
```
nmpc.quad_stage_cost.add_states(names=['x_0', 'x_1', 'x_2'],
                                weights=[100, .1, .1],
                                ref=[0.5, .1, .2])

nmpc.quad_terminal_cost.add_states(names=['x_0', 'x_1', 'x_2'],
                                   weights=np.array([100, .1,  .1]),
                                   ref=[0.5, .1, .2])

```
but you already did this right?

Regarding the plotting, it's done simply with the command

```
nmpc.solution.plot()
```

or do you need the entire code that's under .plot()?

**Reply by wissamkafa** [reply]
Yes, I already did this.
this is correct, when we train the network on a data generated from single setpoint and single initial point,
but when creating a data set from various configurations, this would not be possible.

If there is no direct way in this case to do that, Maybe, after we train the network we use this one more time:
nmpc.quad_stage_cost.add_states(names=['x_0', 'x_1', 'x_2'],
                                weights=[100, .1, .1],
                                ref=[0.5, .1, .2])
nmpc.quad_terminal_cost.add_states(names=['x_0', 'x_1', 'x_2'],
                                   weights=np.array([100, .1,  .1]),
                                   ref=[0.5, .1, .2])
                                
So the network would consider it as the desired setpoint.
do you agree?

**Reply by brunomorampc** [reply]
I will test this and let you know

**Reply by brunomorampc** [reply]
## Modify reference of the learned controller "online"
You are right, you can't modify the reference "online" that easily, since the ANN coming out of that training, if you use our `generate_data` function will take by default only states and inputs as training data. Nevertheless, you can still change the initial conditions (see example below)

If you really need your network to be able to work with difference references, then you need to pass the reference as an extra input to the NN. So the features will be `[ x1_ref, x2_ref, x3_ref, x1,x2,x3]` and the labels the input. 

In this case you need to generate the data yourself, without usingth generate_data. This should be rather easy. Also, you can't use `SimpleControlLoop`  but need to explicitly do the loop yourself like so

```
solution = plant.solution
for step in range(n_steps):
    u = ann.predict(xref, x0_plant)
    plant.simulate(u = u)
    x0_plant = solution['x:f']
```
which is also easy.

## Example: Add new data to database
If instead you want to use the `generate_data` function, and change the initial conditions for every run where you collect the data, you can do like this. Assuming `system` is a Hilo model

```
for k, x_init in enumerate([[0., 0., 0.],[0.1, 0.1, 0.1]]):
    system.set_initial_conditions(x_init)
    # Make controller
    nmpc = NMPC(system)

    # Set horizon
    nmpc.control_horizon = 4
    nmpc.prediction_horizon = 15

    # =============================================================
    # Set cost function
    # =========
    nmpc.quad_stage_cost.add_states(names=['x_0', 'x_1', 'x_2'],
                                    weights=[100, .1, .1],
                                    ref=[0.5, .1, .2])

    nmpc.quad_stage_cost.add_inputs(names=['u_0'], weights=[1])


    nmpc.quad_terminal_cost.add_states(names=['x_0', 'x_1', 'x_2'],
                                       weights=np.array([100, .1,  .1]),
                                       ref=[0.5, .1, .2])

    nmpc.set_box_constraints(x_lb=[-2.2, -1.2, -200], # 1.745 is 100 ded/sec^2
                             x_ub=[2.2, 1.2, 200],    # 4 = 229 deg/sec^2
                             u_lb=-12, u_ub=12)
    # Initial guess
    nmpc.set_initial_guess(x_guess=x_init, u_guess=0)



    # Setup controller
    nmpc.setup(options={'print_level': 1, 'ipopt_debugger': False})
    # ==================================================================================
    print(system)
    print(nmpc)




    # Generate data set
    Tf = 10  # Final time
    n_steps = int(Tf / Ts)
    if k == 0:
        data_set = system.generate_data('closed_loop', nmpc, steps=n_steps, use_input_as_label=True)
    else:
        data_set_new = system.generate_data('closed_loop', nmpc, steps=n_steps, use_input_as_label=True)
        data_set.add_data(data_set_new.raw_data[0],data_set_new.raw_data[1], np.array(data_set_new._raw_data['t']))
        data_set.plot_raw_data()

    # Important to get rid of the solution we had before
    system.reset_solution(keep_initial_conditions=False)

```

I am running the training with the code you sent me using the code in the example to see if it works

**Reply by wissamkafa** [reply]
@brunomorampc 
@BrunoMorabito 
I used this code to generate the data, and train the network.
I think it is able to generate data using different initial conditions and setpoints


sp = [1.57, v,  a]
ic = [ 1.57, 0.0, 0.0]
x0=[1.57, 0, 0]
system.set_initial_conditions(ic)

#Step one 
#First Run to create a dummy dataset, and then build up on it.
#Make controller
nmpc = NMPC(system)
#Set horizon
nmpc.control_horizon = 4
nmpc.prediction_horizon = 15
nmpc.quad_stage_cost.add_states(names=['x_0', 'x_1', 'x_2'], weights=[100, .1, .1], ref=[1.57, v, a])
nmpc.quad_stage_cost.add_inputs(names=['u_0'], weights=[1])
nmpc.quad_terminal_cost.add_states(names=['x_0', 'x_1', 'x_2'], weights=np.array([100, .1,  .1]), ref=[1.57, v, a])
nmpc.set_box_constraints(x_lb=[-2.2, -1.2, -200], x_ub=[2.2, 1.2, 200], u_lb=-12, u_ub=12)
#Initial guess
nmpc.set_initial_guess(x_guess=ic, u_guess=0)
#Setup controller
nmpc.setup(options={'print_level': 0, 'ipopt_debugger': False})
data_set_tot = system.generate_data('closed_loop', nmpc, steps=n_steps, use_input_as_label=True)

#Step 2

#Start the loop to create dataset from various conditions
for ic in tqdm(list_of_initial_conditions, desc ="add_initial_conditions"):
    for sp in list_of_set_points:
        system.set_initial_conditions(ic)
        #Make controller
        nmpc = NMPC(system)
        #Set horizon
        nmpc.control_horizon = 4
        nmpc.prediction_horizon = 15
        nmpc.quad_stage_cost.add_states(names=['x_0', 'x_1', 'x_2'], weights=[100, .1, .1], ref=sp)
        nmpc.quad_stage_cost.add_inputs(names=['u_0'], weights=[1])
        nmpc.quad_terminal_cost.add_states(names=['x_0', 'x_1', 'x_2'], weights=np.array([100, .1,  .1]), ref=sp)
        nmpc.set_box_constraints(x_lb=[-2.2, -1.2, -200], x_ub=[2.2, 1.2, 200], u_lb=-12, u_ub=12)
        #Initial guess
        nmpc.set_initial_guess(x_guess=ic, u_guess=0)

        #Setup controller
        nmpc.setup(options={'print_level': 0, 'ipopt_debugger': False})
        data_set_i = system.generate_data('closed_loop', nmpc, steps=n_steps, use_input_as_label=True)
        data_set_tot.add_data(features=data_set_i.raw_data[0], labels=data_set_i.raw_data[1]) # add data to dataset

=====================Store and Load Dataset=======================================
import pickle
#=========== Store created dataset
'''
with open('gimbal_dataset.pickle', 'wb') as output:
    pickle.dump(data_set_tot, output)
'''
#=========== Load stored dataset

with open('gimbal_dataset.pickle', 'rb') as data:
    data_set_tot = pickle.load(data)

#=========== 
#==================================================================================


#Initialize ANN and create ANN structure
ann = ANN(data_set_tot.features, data_set_tot.labels, learning_rate=1e-3, optimizer='Adam')

#Network Structure now, is chosen to be simple
#ann.add_layers(Layer.dense(100, activation='sigmoid'))
ann.add_layers(Layer.dense(10, activation='ReLU'))
ann.add_layers(Layer.dense(10, activation='ReLU'))
ann.add_layers(Layer.dense(10, activation='ReLU'))

ann.add_layers(Layer.dense(10, activation='ReLU'))
ann.add_layers(Layer.dense(10, activation='ReLU'))
ann.add_layers(Layer.dense(10, activation='ReLU'))

ann.add_layers(Layer.dense(10, activation='ReLU'))
ann.add_layers(Layer.dense(10, activation='ReLU'))
ann.add_layers(Layer.dense(10, activation='ReLU'))



#Add dataset
ann.add_data_set(data_set_tot)

#Setup ANN
ann.setup(device='cpu')

#Train ANN
ann.train(16, 2000, validation_split=.2, patience=50)


#=========================================================
#================= Test the learned controller
ic_test = [1.57, 0, 0]
ic_ann = [1.57, 0, 0]
system.set_initial_conditions(ic_test)

#Create new run to specify the reference setpoint for the learned controller
#=======I am nit sure, if this step will work yet=====================
#Make controller
nmpc = NMPC(system)
#Set horizon
nmpc.control_horizon = 4
nmpc.prediction_horizon = 15
nmpc.quad_stage_cost.add_states(names=['x_0', 'x_1', 'x_2'], weights=[100, .1, .1], ref=[1.57, 0, 0]) #ref=[0.6, 1, 20]
nmpc.quad_stage_cost.add_inputs(names=['u_0'], weights=[1])
nmpc.quad_terminal_cost.add_states(names=['x_0', 'x_1', 'x_2'], weights=np.array([100, .1,  .1]), ref=[1.57, 0, 0]) #ref=[0.6, 1, 20]
nmpc.set_box_constraints(x_lb=[-2.2, -1.2, -200], x_ub=[2.2, 1.2, 200], u_lb=-12, u_ub=12)
#Initial guess
nmpc.set_initial_guess(x_guess=ic_test, u_guess=0)

#Setup controller
nmpc.setup(options={'print_level': 0, 'ipopt_debugger': False})

#=========================================================
#Try out the learned controller
system.reset_solution(keep_initial_conditions=False)
system.set_initial_conditions(ic_ann)
scl = SimpleControlLoop(system, ann)
scl.run(n_steps)
#=======================================
then here would be the ploting code


**I have created a data set of 200 setpoints and a single initial condition (equilibrum point), and now the code is running to tain the network and then I will check the results**

**Comment by wissamkafa** [comment]
did this plot both original nmpc and the learned ann?
"nmpc.solution.plot()"

**Reply by brunomorampc** [reply]
No for both you need to add some extra data just before plotting. Here the code just after you run the close loop simulation

```


# Prepare data for comparison
y_data = []
features, labels = data_set.raw_data
# print(f"Features: {features}\n Labels: {labels}")

# These 2 for loop add the training data
ct = 0
for k, feature in enumerate(data_set.features):
    y_data.append({
        'data': np.append(features[k, :], features[k, -1]),
        'kind': 'line',
        'subplot': ct,
        'label': feature + '_mpc'
    })
    ct += 1

for k, label in enumerate(data_set.labels):
    y_data.append({
        'data': np.append(labels[k, :], labels[k, -1]),
        'kind': 'step',
        'subplot': ct,
        'label': label + '_mpc'
    })
    ct += 1

# Plots
scl.plot(y_data=y_data)
scl.plot()

```

**Comment by wissamkafa** [comment]
Dear Bruno 

@BrunoMorabito @brunomorampc  

After our last discussion, I want to inform you and report the improvement in working with HILO, 

You were right that the training data created by running one simulation is not enough, so 
I created a training dataset with 1000 initial conditions and trained the network to go to a fixed setpoint. 
I used a uniform distribution to randomly choose the initial points from a given interval. 
The results were quite good, and the network closely imitates the NMPC behavior. 

After running hundreds of experiments to choose the best network structure and parameters, I got such good performance, as shown in the photos, below. 

but I have noticed some unusual (or I think so), that I want to share: 
- training on the same dataset, using the same parameters and without changing anything, could give different performances even one training could give a good performance and the other could give a bad performance. I could not figure out the reason for that. 
- when getting a good performance and testing the nn on some initial conditions, it seems that in some random initial conditions, the performance is not as good as in other initial conditions, even the training and evaluation loss is very close to zero like 0.005. 
Can we say here we need a bigger dataset to train and get the same performance for all points? 
- Training takes a long time, so each modification took me quite a long time to test. 
- I managed to save the trained network and reload it, so now I can save it and test it on other PCs, or for comparison purposes. (I used torch.save(ann)), but the size of the saved file is 32MB and I want to ask, if this is just the network or if it could save something else? 
- Still, plotting the NMPC and ann behavior on the same plot, I could not figure out how to do it yet.
- In the next steps, I will involve the setpoint as a feature in the network and train it for variable setpoint. 

For Now, I'm trying to simulate a more complex system (the first one was a simplified version), the system I have is of the form: 
    dx_0 = x_1 
    dx_1 = a1*x_1*x_4 + a2*x_1 + a3*x_2 
    dx_2 = b1*x_2 + b2*x_1 + b3*u_0 
    dx_3 = x_4 
    dx_4 = c1*x_1*x_4 + c2*x_4 + c3*x_5 
    dx_5 = d1*x_5 + d2*x_4 + d3*u_1 
some of a1, a2, a3, b1, b2, b3, c1, c2, c3, d1, d2, d3  include sin and cosine of x_1 and x_3. 

The problem here is when trying to create the NMPC with HILO, I always get "infeasible_problem_detected". 
I tried to:  
- use several initial guess points 
- relax the constraints 
- change the weights in stage and terminal cost 

but always I get the same error  "infeasible_problem_detected". 
I'm just wondering if HILO can solve such a system. 
I can't find a reason to say that the problem itself is infeasible, so, can you please advise me here? 



![ic_test_-2 0](https://github.com/hilo-mpc/hilo-mpc/assets/9342111/ec6477b9-3e8c-4f1f-bb62-d2f5136e96bf)
![ic_test_-2 0_2nd](https://github.com/hilo-mpc/hilo-mpc/assets/9342111/ca3c0fef-fa86-492c-8f0b-5b39b62caa91)
![Training and Evaluation Loss_2nd](https://github.com/hilo-mpc/hilo-mpc/assets/9342111/5ae33a14-ab4f-416f-946e-53ec411095bd)
